### DISCUSSION TITLE
DS18x20

### DISCUSSION THREAD

**Discussion by cretuchris** [arendst/Tasmota#13293] [Q&A] [answered]
Dear all, first of all Iwould like to thank you for the great job you are doing here, keep it up like this!

second I would like to ask for your opinion/help:
I have 8 DS18B20 sensors connected to a ESP32 flashed with the newest Tasmota software,
Q1: is it not possible to select more DS18x20 inputs on the Tasmota/ESP ?
when I select a second one no sensoer will be recognized (whit and without pull/up resistor)

Q2:every time I power down or atach a new sensor to the Tasmota/ESP it will be shown with a different ID number.
is it not possible to fix the ID number with the serial number of the sensor somehow ?

I already searched the internet for an answer but I was not able to find any.

Thanks in advance for your answers.

**Comment by sfromis** [comment, marked as answer]
Q1: Only one pin, but you can get more than 8 sensors using `#define DS18X20_MAX_SENSORS  8`
Q2: You cannot choose the number, but you can indeed replace using `#define DS18x20_USE_ID_AS_NAME`
Both are settings at build time. You could have found them in the [docs](https://tasmota.github.io/docs/DS18x20/).

**Reply by cretuchris** [reply]
Thanks. I’ll try this. And instead of ID the sensor id, and instead of NAME the name I’d like to give the Sensor ?
He will also send the name over MQTT ?

**Reply by sfromis** [reply]
You should have read the docs page I linked. It is a compile-time option, exactly as written, no replacements. You cannot select a name, it is basically what you asked for.
*Use last 3 bytes for naming of sensors*

And yes, the changed name will also be used as sensor name in the JSON published to MQTT.

**Reply by cretuchris** [reply]
I understand. And afterwards it will not change the ID anymore ?
I’m getting the actual ID over MQTT
My question/problem is if I restart the ESP32 he will give the ID’s again to different Sensors/serial number …
Or if I have to reflash it because of a mistake (p.s. I’m a beginner)

thanks

**Reply by sfromis** [reply]
Assuming that by `ID` you mean the `Id` field included in the JSON object for each sensor, this is from the sensor hardware, outside of what Tasmota has control over. It would be the same moving the sensor to a different board, as it is basically a fixed serial number for each sensor. If you get two with the same, the manufacturer has screwed up. With the mentioned build option set, the last part of this will be used in the JSON object name instead of the default 1-digit sequence number.

**Reply by barbudor** [reply]
Serial number is burnt into the DS18x20 chip itself. For a given chip it will not change.

**Reply by cretuchris** [reply]
the serial number is alway different.
please see the atached picture.
Tasmota_ESP sends an ID with a specific unique Serial number.
each time I restart the Tasmota or reflash it it will change the ID given to the serial number, this way it is very hard to use it like this for an example too control a pump with a script.


![Tasmota_ID_SERIAL](https://user-images.githubusercontent.com/82668674/136492033-682cfc22-6af2-4d25-9675-86b944f30557.png)

![Tasmota](https://user-images.githubusercontent.com/82668674/136492072-494b72b5-f282-4b29-921f-9fc1f55731de.png)

**Reply by barbudor** [reply]
@cretuchris You are just not ready the docs correctly as you ARE NOT USING the chip's serial number
 Did you followed the compile instructions ? 
https://tasmota.github.io/docs/DS18x20/#compile-options
and https://tasmota.github.io/docs/Compile-your-build/

**Reply by sfromis** [reply]
To be clear about terms.... 

The "serial number" is the long hex number for each DS18x20 sensor, like `3C01D075EF00`, which Tasmota puts into the field `Id`. This value never changes.

When there are more than one DS18x20 sensor, Tasmota will by default add a number (1-8) to the sensor name getting you to a name like `DS18B20-1`. While this works to make different names, this short number can and will indeed change depending on how many and which sensors are connected, and thus is not useful for knowing which physical sensor it is.

This was not very practical, and a build time option was added to replace that short sequence number by a longer code from the last 3 bytes of the hardware serial number. This would give something like `DS18B20-75EF00` instead of `DS18B20-1`. Basically just what you asked for.

To do this, you *must* compile your own binary where you set the compile option (and maybe also increase the max number of connected sensors). It is not at all difficult to do, but you do need a little bit of reading to get started.

**Comment by cretuchris** [comment]
now I understand. I have to modify the source code.
thank you all for the help.
Have a nice weekend!

**Comment by cretuchris** [comment]
Dear all. I’ve managed to do the change you mentioned and it’s working fine. But I still have one small problem. I am reading the value sent over MQTT, and I select for example the position of the sensor not the name, if I re boot again or I attach another sensor, the position of the sensor will change again. This is why I asked if it’s possible somehow to fix the sensor to a position ?
Ps. Sometimes I can see on the ESP that a sensor disappears for a few seconds and then his position will be taken by another until it comes back. Which is also a problem …

**Reply by sfromis** [reply]
If by "position" you mean the same senor to always be number 3 in the list, irrespective of name, no, that cannot be made static if number of  available sensors vary. Just do not use such a bad method, when using the option to fix the name, you always get the same name, which is excellent for extracting the values. Trying to use the relative position is just doing it wrong :-)
